nvme_tcp.go

```diff
@@ -0,0 +1,658 @@
+		}
+	}
+}
+
+func (c *NVMeTCPConnector) cleanConnection(ctx context.Context, force bool, info NVMeTCPVolumeInfo) error {
+	defer tracer.TraceFuncCall(ctx, "NVMeTCPConnector.cleanConnection")()
+	var devices []string
+	wwn := info.WWN
+
+	namespaceDevices := c.nvmeTCPLib.ListNamespaceDevices()
+
+	for devicePath, _ := range namespaceDevices {
+		for _, namespace := range namespaceDevices[devicePath] {
+			nguid, _ := c.nvmeTCPLib.GetNamespaceData(devicePath, namespace)
+			if c.wwnMatches(nguid, wwn) {
+				devices = append(devices, devicePath)
+			}
+		}
+	}
+	if len(devices) == 0 {
+		return nil
+	}
+	return c.baseConnector.cleanDevices(ctx, force, devices)
+}
+
+func (c *NVMeTCPConnector) connectSingleDevice(
+	ctx context.Context, sessions []goiscsi.ISCSISession, info ISCSIVolumeInfo) (Device, error) {
+	defer tracer.TraceFuncCall(ctx, "NVMeTCPConnector.connectSingleDevice")()
+	devCH := make(chan string, len(sessions))
+	wg := sync.WaitGroup{}
+	//discoveryCtx, cFunc := context.WithTimeout(ctx, c.waitDeviceTimeout)
+	_, cFunc := context.WithTimeout(ctx, c.waitDeviceTimeout)
+	defer cFunc()
+
+	/*for _, s := range sessions {
+		wg.Add(1)
+		//go c.discoverDevice(discoveryCtx, &wg, devCH, s, info)
+	}*/
+	// for non blocking wg wait
+	wgCH := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(wgCH)
+	}()
+
+	var devices []string
+	var wwn string
+	var discoveryComplete, lastTry bool
+	var endTime time.Time
+	for {
+		// get discovered devices
+		select {
+		case <-ctx.Done():
+			return Device{}, errors.New("connectSingleDevice canceled")
+		default:
+		}
+		devices = readDevicesFromResultCH(devCH, devices)
+		// check all discovery gorutines finished
+		if !discoveryComplete {
+			select {
+			case <-wgCH:
+				discoveryComplete = true
+				logger.Info(ctx, "all discovery goroutines complete")
+			default:
+				logger.Info(ctx, "discovery goroutines are still running")
+			}
+		}
+		if discoveryComplete && len(devices) == 0 {
+			msg := "discovery complete but devices not found"
+			logger.Error(ctx, msg)
+			return Device{}, errors.New(msg)
+		}
+		if wwn == "" && len(devices) != 0 {
+			var err error
+			wwn, err = c.scsi.GetDeviceWWN(ctx, devices)
+			if err != nil {
+				logger.Info(ctx, "wwn for devices %s not found", devices)
+			}
+		}
+		if wwn != "" {
+			for _, d := range devices {
+				if err := c.scsi.WaitUdevSymlinkNVMe(ctx, d, wwn); err == nil {
+					logger.Info(ctx, "registered device found: %s", d)
+					return Device{Name: d, WWN: wwn}, nil
+				}
+			}
+		}
+		if discoveryComplete && !lastTry {
+			logger.Info(ctx, "discovery finished, wait %f seconds for device registration",
+				c.waitDeviceRegisterTimeout.Seconds())
+			lastTry = true
+			endTime = time.Now().Add(c.waitDeviceRegisterTimeout)
+		}
+		if lastTry && time.Now().After(endTime) {
+			msg := "registered device not found"
+			logger.Error(ctx, msg)
+			return Device{}, errors.New(msg)
+		}
+		time.Sleep(time.Second)
+	}
+}
+
+//inprogress
```

Comment:
Please add the appropriate comment for this function

Reply:
changed function comments